src/kafunk/Utility/Binary.fs

```diff
@@ -402,45 +339,48 @@ type BinaryZipper (buf:ArraySegment<byte>) =
     let value = go 0 0
     (value >>> 1) ^^^ -(value &&& 1)
 
+  member __.WriteVarint (value:int) =
+    let mutable v = (value <<< 1) ^^^ (value >>> 31)
+    while ((v &&& 0xffffff80) <> 0) do
+      let b = byte ((v &&& 0x7f) ||| 0x80)
+      __.WriteByte b
+      v <- v >>> 7
+    __.WriteByte (byte v)
+    
+  member __.WriteVarint64 (value:int64) =
+    let mutable v = (value <<< 1) ^^^ (value >>> 63)
+    while ((v &&& 0xffffffffffffff80L) <> 0L) do
+      let b = byte ((v &&& 0x7fL) ||| 0x80L)
+      __.WriteByte b
+      v <- v >>> 7
+    __.WriteByte (byte v)
+
+  member __.ReadVarint64 () : int64 =
+    let rec go value i =
+      let b = int <| __.ReadInt8 ()
+      if (b &&& 0x80 <> 0) then
+        let value = value ||| int64 ((b &&& 0x7f) <<< i)
+        let i = i + 7
+        if (i > 63) then failwith "invalid varint64" 
+        else go value i
+      else
+        value ||| int64 (b <<< i)
+    let value = go 0L 0
+    (value >>> 1) ^^^ -(value &&& 1L)
```

Comment:
Ah okay I think I see what I did wrong now. Kafka was never "doubling" anything, it was just zigzag encoding a signed int64. Sorry about that!